_posts/2024-10-24-autorebalancing-on-scaling.md

```diff
@@ -0,0 +1,150 @@
+
+While both the above approaches work, the process is still actually manual and done in two separate steps:
+
+* scale the cluster up, then run a rebalancing.
+* run a rebalancing, then scale the cluster down.
+
+What if it could be possible to automate such a process?
+What about having the Strimzi operator running a rebalance right after the scale up operation, or right before the attempt of scaling down the cluster?
+The new Strimzi 0.44.0 release brings to you the new auto-rebalancing on cluster scaling feature!
+
+### The auto-rebalancing on scaling feature to the rescue!
+
+Enabling the auto-rebalancing is totally configurable through the new `autoRebalance` property within the `cruiseControl` section in the `Kafka` custom resource.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      - mode: add-brokers
+        template:
+          name: my-add-brokers-rebalancing-template
+      - mode: remove-brokers
+        template:
+          name: my-remove-brokers-rebalancing-template
+```
+
+From the example above, within the `autoRebalance` section, the user can specify "when" the auto-rebalancing has to run: on scaling up (by using the `add-brokers` mode) and/or on scaling down (by using the `remove-broker` mode).
+It is not mandatory to have them both; you can decide to have auto-rebalancing on scaling up but not on scaling down and vice versa.
+For each mode, it is possible to specify the name of a rebalancing configuration "template".
+This is just a `KafkaRebalance` custom resource with the new `strimzi.io/rebalance-template: true` annotation applied, which makes it a configuration template to be used during auto-rebalancing operations and not an actual rebalance request to run.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance-template
+  annotations:
+    strimzi.io/rebalance-template: "true" # specifies that this KafkaRebalance is a rebalance configuration template
+spec:
+  # NOTE: mode and brokers fields, if set, will be just ignored because they are
+  #       automatically set on the corresponding KafkaRebalance by the operator
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+The configuration template can be the same for both scaling operations or the user can decide to use different ones.
+Using the `KafkaRebalance` custom resource is a way to make lives easier to the Strimzi users because they already know how it works in regards the integration with Cruise Control.
```

Comment:
```suggestion
Using the `KafkaRebalance` custom resource is a way to make the lives of Strimzi users easier because they already know how it works in regards to integration with Cruise Control.
```